File: Durian/Tools/unreal_mcp_server.js
const readline = require('readline');
const http = require('http');

const base = 'http://127.0.0.1:30010';
const tools = [
  { name: 'unreal_status', description: 'Check Unreal Remote Control API.', inputSchema: { type: 'object', properties: {} } },
  { name: 'unreal_get_preset', description: 'Read an Unreal Remote Control Preset.', inputSchema: { type: 'object', properties: { preset: { type: 'string' } }, required: ['preset'] } },
  { name: 'unreal_set_property', description: 'Set an exposed Unreal property.', inputSchema: { type: 'object', properties: { object_path: { type: 'string' }, property_name: { type: 'string' }, property_value: {} }, required: ['object_path', 'property_name', 'property_value'] } }
];

function request(method, path, body) {
  return new Promise((resolve, reject) => {
    const data = body === undefined ? null : Buffer.from(JSON.stringify(body));
    const req = http.request(base + path, { method, headers: data ? { 'Content-Type': 'application/json', 'Content-Length': data.length } : {} }, res => {
      let raw = ''; res.setEncoding('utf8');
      res.on('data', x => raw += x);
      res.on('end', () => { try { resolve(JSON.parse(raw)); } catch { resolve(raw); } });
    });
    req.on('error', reject); req.setTimeout(5000, () => req.destroy(new Error('Unreal API timeout')));
    if (data) req.write(data); req.end();
  });
}
function out(id, result) { process.stdout.write(JSON.stringify({ jsonrpc: '2.0', id, result }) + '\n'); }
async function handle(m) {
  if (m.method === 'initialize') return out(m.id, { protocolVersion: '2024-11-05', capabilities: { tools: {} }, serverInfo: { name: 'durian-unreal', version: '0.1.0' } });
  if (m.method === 'tools/list') return out(m.id, { tools });
  if (m.method === 'notifications/initialized') return;
  if (m.method !== 'tools/call') return out(m.id, {});
  const a = m.params.arguments || {}, name = m.params.name;
  try {
    let value;
    if (name === 'unreal_status') value = await request('GET', '/remote/info');
    else if (name === 'unreal_get_preset') value = await request('GET', '/remote/preset/' + encodeURIComponent(a.preset));
    else if (name === 'unreal_set_property') value = await request('PUT', '/remote/object/property', { ObjectPath: a.object_path, PropertyName: a.property_name, PropertyValue: a.property_value });
    else throw new Error('Unknown tool: ' + name);
    out(m.id, { content: [{ type: 'text', text: JSON.stringify(value, null, 2) }] });
  } catch (e) { out(m.id, { isError: true, content: [{ type: 'text', text: e.message }] }); }
}
readline.createInterface({ input: process.stdin }).on('line', line => { try { handle(JSON.parse(line)); } catch (e) {} });

File: Durian/Tools/unreal_mcp_server.py
"""Small stdio MCP bridge for Unreal Engine Remote Control API.

This intentionally uses only the Python standard library.
"""
import json
import sys
import urllib.error
import urllib.parse
import urllib.request

BASE = "http://127.0.0.1:30010"


def http(method, path, body=None):
    data = None if body is None else json.dumps(body).encode("utf-8")
    req = urllib.request.Request(
        BASE + path,
        data=data,
        method=method,
        headers={"Content-Type": "application/json"},
    )
    with urllib.request.urlopen(req, timeout=5) as response:
        raw = response.read().decode("utf-8")
        try:
            return json.loads(raw)
        except json.JSONDecodeError:
            return raw


TOOLS = [
    {
        "name": "unreal_status",
        "description": "Check whether Unreal Engine Remote Control API is reachable.",
        "inputSchema": {"type": "object", "properties": {}},
    },
    {
        "name": "unreal_get_preset",
        "description": "Read a Remote Control Preset from Unreal Engine.",
        "inputSchema": {
            "type": "object",
            "properties": {"preset": {"type": "string"}},
            "required": ["preset"],
        },
    },
    {
        "name": "unreal_set_property",
        "description": "Set an exposed Unreal object property through Remote Control.",
        "inputSchema": {
            "type": "object",
            "properties": {
                "object_path": {"type": "string"},
                "property_name": {"type": "string"},
                "property_value": {},
            },
            "required": ["object_path", "property_name", "property_value"],
        },
    },
]


def result(value):
    return {"content": [{"type": "text", "text": json.dumps(value, ensure_ascii=False, indent=2)}]}


def handle(message):
    method = message.get("method")
    params = message.get("params") or {}
    if method == "initialize":
        return {
            "protocolVersion": "2024-11-05",
            "capabilities": {"tools": {}},
            "serverInfo": {"name": "durian-unreal", "version": "0.1.0"},
        }
    if method == "notifications/initialized":
        return None
    if method == "tools/list":
        return {"tools": TOOLS}
    if method != "tools/call":
        return {}

    name = params.get("name")
    args = params.get("arguments") or {}
    try:
        if name == "unreal_status":
            return result(http("GET", "/remote/info"))
        if name == "unreal_get_preset":
            preset = urllib.parse.quote(args["preset"], safe="")
            return result(http("GET", f"/remote/preset/{preset}"))
        if name == "unreal_set_property":
            body = {
                "ObjectPath": args["object_path"],
                "PropertyName": args["property_name"],
                "PropertyValue": args["property_value"],
            }
            return result(http("PUT", "/remote/object/property", body))
        raise ValueError(f"Unknown tool: {name}")
    except Exception as exc:
        return {"isError": True, "content": [{"type": "text", "text": str(exc)}]}


for line in sys.stdin:
    try:
        message = json.loads(line)
        response = handle(message)
        if "id" in message and response is not None:
            print(json.dumps({"jsonrpc": "2.0", "id": message["id"], "result": response}), flush=True)
    except Exception as exc:
        if "id" in locals().get("message", {}):
            print(json.dumps({"jsonrpc": "2.0", "id": message["id"], "error": {"code": -32000, "message": str(exc)}}), flush=True)
